hack/release.sh

```diff
@@ -0,0 +1,219 @@
+#!/bin/bash
+
+# Copyright 2019 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script is used build new container images of the CAPV manager and
+# clusterctl. When invoked without arguments, the default behavior is to build
+# a new image for the manager and to tag it as `edge`
+#
+# Assumptions:
+#  If using -p to push images, then the env var GCR_KEY_FILE must be set
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# BASE_REPO is the root path of the image repository
+readonly BASE_IMAGE_REPO=gcr.io/cluster-api-provider-vsphere
+
+# Release images
+readonly CAPV_MANAGER_IMAGE_RELEASE=${BASE_IMAGE_REPO}/release/manager
+readonly CAPV_MANIFESTS_IMAGE_RELEASE=${BASE_IMAGE_REPO}/release/manifests
+
+# PR images
+readonly CAPV_MANAGER_IMAGE_PR=${BASE_IMAGE_REPO}/pr/manager
+readonly CAPV_MANIFESTS_IMAGE_PR=${BASE_IMAGE_REPO}/pr/manifests
+
+# CI images
+readonly CAPV_MANAGER_IMAGE_CI=${BASE_IMAGE_REPO}/ci/manager
+readonly CAPV_MANIFESTS_IMAGE_CI=${BASE_IMAGE_REPO}/ci/manifests
+
+PUSH=
+PR=
+MANAGER_IMAGE_NAME=
+MANIFESTS_IMAGE_NAME=
+VERSION=
+LATEST=
+RELEASE=
+GCR_KEY_FILE=${GCR_KEY_FILE:-}
+
+# Change directories to the parent directory of the one in which this
+# script is located.
+cd "$(dirname "${BASH_SOURCE[0]}")/.."
+
+function usage() {
+  cat <<EOF
+usage: ${0} [FLAGS]
+  Builds and optionally pushes new images for Cluster API Provider vSphere (CAPV)
+
+FLAGS
+  -h    Show this help and exit
+  -k    Path to GCR key file. Overriden if GCR_KEY_FILE env var is set
+  -l    tag the images as "latest" in addition to their version
+        when used with -p, both tags will be pushed
+  -p    push the images to the public container registry
+  -t    the images are solely for the use by CI/testing
```

Comment:
Hi @codenrhoden,

I'm wondering if this should stand for `type` of build, and have its valid values be `ci|release|pr`. That would enable `case/esac` for the build type:

```shell
case "${BUILD_RELEASE_TYPE}" in
  ci)
    ;;
  release)
    ;;
  pr)
    ;;
esac
```

So the changes might look like:

```shell
BUILD_RELEASE_TYPE ="${BUILD_RELEASE_TYPE-}"

# If BUILD_RELEASE_TYPE is not set then check to see if this is a PR
# release build. This may still be overridden below with the "-t" flag.
if [ -z "${BUILD_RELEASE_TYPE}" ]; then
  if hack/match-release-tag.sh >/dev/null 2>&1; then
    BUILD_RELEASE_TYPE=release
  else
    BUILD_RELEASE_TYPE=pr
  fi
fi

...

function usage() {
  cat <<EOF
usage: ${0} [FLAGS]
  Builds and optionally pushes new images for Cluster API Provider vSphere (CAPV)
FLAGS
  -h    show this help and exit
  -k    path to GCR key file. Overriden if GCR_KEY_FILE env var is set
  -l    tag the images as "latest" in addition to their version
        when used with -p, both tags will be pushed
  -p    push the images to the public container registry
  -t    the build/release type (defaults to ${BUILD_RELEASE_TYPE})
EOF
}

...

# Start of main script
while getopts ":hk:lpt:" opt; do
  case ${opt} in
    h)
      usage 1>&2; exit 1
      ;;
    k)
      GCR_KEY_FILE="${OPTARG}"
      ;;
    l)
      LATEST=1
      ;;
    p)
      PUSH=1
      ;;
    t)
      BUILD_RELEASE_TYPE="${OPTARG}"
      ;;
    \?)
      { echo "invalid option: -${OPTARG}"; usage; } 1>&2; exit 1
      ;;
    :)
      echo "option -${OPTARG} requires an argument" 1>&2; exit 1
      ;;
  esac
done
shift $((OPTIND-1))

# Verify the GCR_KEY_FILE exists
[ -e "${GCR_KEY_FILE}" ] || fatal "key file ${GCR_KEY_FILE} does not exist"

# Validate build/release type.
case "${BUILD_RELEASE_TYPE}" in
  ci|pr|release)
    # do nothing
    ;;
  *)
    fatal "invalid BUILD_RELEASE_TYPE: ${BUILD_RELEASE_TYPE}"
    ;;
esac
```